ssh/tailssh/tailssh.go

```diff
@@ -253,42 +272,46 @@ func toIPPort(a net.Addr) (ipp netaddr.IPPort) {
 }
 
-// evaluatePolicy returns the SSHAction, sshConnInfo and localUser after
-// evaluating the sshUser and remoteAddr against the SSHPolicy. The remoteAddr
-// and localAddr params must be Tailscale IPs.
-//
-// The return sshConnInfo will be non-nil, even on some errors, if the
-// evaluation made it far enough to resolve the remoteAddr to a Tailscale IP.
-func (srv *server) evaluatePolicy(sshUser string, localAddr, remoteAddr netaddr.IPPort, pubKey ssh.PublicKey) (_ *tailcfg.SSHAction, _ *sshConnInfo, localUser string, _ error) {
-	pol, ok := srv.sshPolicy()
-	if !ok {
-		return nil, nil, "", fmt.Errorf("tailssh: rejecting connection; no SSH policy")
+// connInfo returns a populated sshConnInfo from the provided arguments,
+// validating only that they represent a known Tailscale identity.
+func (srv *server) connInfo(sshUser string, localAddr, remoteAddr netaddr.IPPort) (*sshConnInfo, error) {
+	ci := &sshConnInfo{
+		now:                srv.now(),
+		fetchPublicKeysURL: srv.fetchPublicKeysURL,
+		sshUser:            sshUser,
+		src:                remoteAddr,
+		dst:                localAddr,
 	}
 	if !tsaddr.IsTailscaleIP(remoteAddr.IP()) {
-		return nil, nil, "", fmt.Errorf("tailssh: rejecting non-Tailscale remote address %v", remoteAddr)
+		return ci, fmt.Errorf("tailssh: rejecting non-Tailscale remote address %v", remoteAddr)
 	}
 	if !tsaddr.IsTailscaleIP(localAddr.IP()) {
-		return nil, nil, "", fmt.Errorf("tailssh: rejecting non-Tailscale remote address %v", localAddr)
+		return ci, fmt.Errorf("tailssh: rejecting non-Tailscale remote address %v", localAddr)
 	}
 	node, uprof, ok := srv.lb.WhoIs(remoteAddr)
 	if !ok {
-		return nil, nil, "", fmt.Errorf("unknown Tailscale identity from src %v", remoteAddr)
+		return ci, fmt.Errorf("unknown Tailscale identity from src %v", remoteAddr)
 	}
-	ci := &sshConnInfo{
-		now:                srv.now(),
-		fetchPublicKeysURL: srv.fetchPublicKeysURL,
-		sshUser:            sshUser,
-		src:                remoteAddr,
-		dst:                localAddr,
-		node:               node,
-		uprof:              &uprof,
-		pubKey:             pubKey,
+
+	ci.node = node
+	ci.uprof = &uprof
+	return ci, nil
+}
+
+// evaluatePolicy returns the SSHAction, sshConnInfo and localUser after
+// evaluating the sshUser and remoteAddr against the SSHPolicy. The remoteAddr
+// and localAddr params must be Tailscale IPs. The pubKey may be nil for "none"
+// auth.
+func (srv *server) evaluatePolicy(ci *sshConnInfo, pubKey gossh.PublicKey) (_ *tailcfg.SSHAction, localUser string, _ error) {
+	pol, ok := srv.sshPolicy()
+	if !ok {
+		return nil, "", fmt.Errorf("tailssh: rejecting connection; no SSH policy")
 	}
-	a, localUser, ok := evalSSHPolicy(pol, ci)
+	a, localUser, ok := srv.evalSSHPolicy(pol, ci, pubKey)
 	if !ok {
-		return nil, ci, "", fmt.Errorf("ssh: access denied for %q from %v", uprof.LoginName, ci.src.IP())
+		return nil, "", fmt.Errorf("ssh: access denied for %q from %v", ci.uprof.LoginName, ci.src.IP())
```

Comment:
it's not so much that access was denied as there was no matching access grant.  This is a little misleading because the following line isn't the opposite of "access denied": line 314 could return an SSHAction that's Rejected.